File: modules/local-platform/android/src/main/java/expo/modules/localplatform/LocalManagedFiles.kt
package expo.modules.localplatform

import android.content.Context
import java.io.File
import java.net.URI

internal object LocalManagedFiles {
  private fun inbox(c: Context) = File(c.noBackupFilesDir, "sns-gateway/inbox").canonicalFile
  private fun staging(c: Context) = File(c.cacheDir, "sns-gateway/share").canonicalFile
  private fun permitted(c: Context, raw: String): File {
    val uri = URI(raw)
    require(uri.scheme == "file" && uri.authority.isNullOrEmpty() && uri.query == null && uri.fragment == null)
    val file = File(uri).canonicalFile
    val inInbox = file.parentFile == inbox(c) && file.name.matches(Regex("[a-f0-9]{64}\\.jpg"))
    val inStage = file.parentFile?.parentFile == staging(c) && file.parentFile!!.name.matches(Regex("[A-Fa-f0-9-]{36}")) && file.name.matches(Regex("(?:fixture-)?[0-9]+\\.jpg"))
    require(inInbox || inStage) { "UNMANAGED_FILE" }
    require(!file.exists() || file.isFile) { "NOT_A_FILE" }
    return file
  }
  fun inventory(c: Context): List<Map<String, Any>> {
    val files = inbox(c).listFiles().orEmpty().toList() + staging(c).listFiles().orEmpty().filter { it.isDirectory }.flatMap { it.listFiles().orEmpty().toList() }
    return files.mapNotNull { candidate ->
      try {
        val file = permitted(c, candidate.toURI().toString())
        if (!file.isFile) null else mapOf("uri" to file.toURI().toString(), "kind" to if (file.parentFile == inbox(c)) "INBOX" else "STAGING", "bytes" to file.length(), "modifiedAt" to file.lastModified())
      } catch (_: Exception) { null }
    }
  }
  fun requireSpace(c: Context, extra: Long) {
    require(extra >= 0 && inventory(c).sumOf { (it["bytes"] as Number).toLong() } + extra <= 524_288_000) { "LOCAL_STORAGE_FULL" }
  }
  fun delete(c: Context, uris: List<String>): List<String> {
    require(uris.size <= 100 && uris.distinct().size == uris.size)
    val files = uris.map { permitted(c, it) } // Validate all paths before deleting any file.
    return uris.zip(files).mapNotNull { (uri, file) ->
      if (!file.exists() || file.delete()) {
        val parent = file.parentFile
        if (parent?.parentFile == staging(c) && parent.listFiles()?.isEmpty() == true) parent.delete()
        uri
      } else null
    }
  }
}

File: modules/local-platform/ios/LocalManagedFiles.swift
import Foundation

final class LocalManagedFiles {
  static func support() throws -> URL {
    var url = try FileManager.default.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true).appendingPathComponent("sns-gateway", isDirectory: true)
    try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
    var values = URLResourceValues(); values.isExcludedFromBackup = true
    try url.setResourceValues(values)
    return url.standardizedFileURL.resolvingSymlinksInPath()
  }
  static func shareRoot() throws -> URL {
    try FileManager.default.url(for: .cachesDirectory, in: .userDomainMask, appropriateFor: nil, create: true).appendingPathComponent("sns-gateway/share", isDirectory: true).standardizedFileURL.resolvingSymlinksInPath()
  }
  private static func allowed(_ raw: String) throws -> URL {
    guard let uri = URL(string: raw), uri.isFileURL, uri.host == nil || uri.host == "", uri.query == nil, uri.fragment == nil else { throw NSError(domain: "UnmanagedFile", code: 1) }
    let file = uri.standardizedFileURL.resolvingSymlinksInPath()
    let inbox = try support().appendingPathComponent("inbox", isDirectory: true)
    let parent = file.deletingLastPathComponent()
    let inboxFile = parent == inbox && file.lastPathComponent.range(of: "^[a-f0-9]{64}\\.jpg$", options: .regularExpression) != nil
    let sharedRoot = try shareRoot()
    let staged = parent.deletingLastPathComponent() == sharedRoot && UUID(uuidString: parent.lastPathComponent) != nil
      && file.lastPathComponent.range(of: "^(fixture-)?[0-9]+\\.jpg$", options: .regularExpression) != nil
    guard inboxFile || staged else { throw NSError(domain: "UnmanagedFile", code: 2) }
    if FileManager.default.fileExists(atPath: file.path) {
      guard try file.resourceValues(forKeys: [.isRegularFileKey]).isRegularFile == true else { throw NSError(domain: "UnmanagedFile", code: 3) }
    }
    return file
  }
  static func inventory() throws -> [[String: Any]] {
    let inbox = try support().appendingPathComponent("inbox", isDirectory: true)
    let share = try shareRoot()
    let original = (try? FileManager.default.contentsOfDirectory(at: inbox, includingPropertiesForKeys: nil)) ?? []
    let folders = (try? FileManager.default.contentsOfDirectory(at: share, includingPropertiesForKeys: nil)) ?? []
    let files = original + folders.flatMap { (try? FileManager.default.contentsOfDirectory(at: $0, includingPropertiesForKeys: nil)) ?? [] }
    return files.compactMap { candidate in
      guard let file = try? allowed(candidate.absoluteString),
        let values = try? file.resourceValues(forKeys: [.fileSizeKey, .contentModificationDateKey, .isRegularFileKey]), values.isRegularFile == true else { return nil }
      return ["uri": file.absoluteString, "kind": file.deletingLastPathComponent() == inbox ? "INBOX" : "STAGING",
        "bytes": values.fileSize ?? 0, "modifiedAt": Int64((values.contentModificationDate?.timeIntervalSince1970 ?? 0) * 1000)]
    }
  }
  static func requireSpace(_ extra: Int) throws {
    let used = try inventory().reduce(0) { $0 + ($1["bytes"] as? Int ?? 0) }
    guard extra >= 0 && used + extra <= 524_288_000 else { throw NSError(domain: "LocalStorageFull", code: 1) }
  }
  static func delete(_ uris: [String]) throws -> [String] {
    guard uris.count <= 100, Set(uris).count == uris.count else { throw NSError(domain: "DeleteLimit", code: 1) }
    let files = try uris.map { try allowed($0) } // Validate the entire request before any mutation.
    var removed = [String]()
    for (raw, file) in zip(uris, files) {
      do {
        if FileManager.default.fileExists(atPath: file.path) { try FileManager.default.removeItem(at: file) }
        removed.append(raw)
        let parent = file.deletingLastPathComponent()
        if parent.deletingLastPathComponent() == (try shareRoot()), (try? FileManager.default.contentsOfDirectory(atPath: parent.path).isEmpty) == true {
          try? FileManager.default.removeItem(at: parent)
        }
      } catch { /* Keep deletion journal pending; never delete originals or the database. */ }
    }
    return removed
  }
}
